pennylane/gradients/parameter_shift.py

```diff
@@ -140,15 +140,36 @@ def _extract_unshifted(recipe, at_least_one_unshifted, f0, gradient_tapes, tape)
     return recipe, at_least_one_unshifted, unshifted_coeff
 
 
-def _unshifted_coeff(g, unshifted_coeff, r0):
-    """Auxiliary function; if unshifted term exists, add its contribution."""
+def _single_meas_grad(result, coeffs, unshifted_coeff, r0):
+    """Compute the gradient for a single measurement by taking the linear combination of the coefficients and the
+    measurement result.
+
+    If an unshifted term exists, its contribution is added to the gradient.
+    """
+    coeffs = qml.math.convert_like(coeffs, result)
+    g = qml.math.tensordot(result, coeffs, [[0], [0]])
     if unshifted_coeff is not None:
         # add the unshifted term
         g = g + unshifted_coeff * r0
+        g = qml.math.array(g)

```

Comment:
No idea which would be preferred. Frankly, this is easier because `g` may be non-tensor-like due to either `g`, `unshifted_coeff`, `r0`.

I can investigate more if we think this is important to track down.